src/ornament.js

```diff
@@ -27,27 +27,100 @@ export class Ornament extends Modifier {
   static format(ornaments, state) {
     if (!ornaments || ornaments.length === 0) return false;
 
-    let width = 0;
+    let width = 0;  // width is used by ornaments, which are always centered on the note head
+    let right_shift = state.right_shift;  // jazz ornaments calculate r/l shift separately
+    let left_shift = state.left_shift;
+    let yOffset = 0;
+
     for (let i = 0; i < ornaments.length; ++i) {
       const ornament = ornaments[i];
       const increment = 2;
 
-      width = Math.max(ornament.getWidth(), width);
+      // Jazz ornaments have their own metrics, and their position depends generally on
+      // the ornament.  So we calculate their position differently.
+      if (ornament.jazzMetrics) {
+        const reportedWidth = ornament.metrics.reportedWidth;
+        if (Ornament.ornamentRelease.indexOf(ornament.type) >= 0) {
+          ornament.render_options.xOffset += (right_shift + 2);
+        }
+        if (Ornament.ornamentAttack.indexOf(ornament.type) >= 0) {
+          ornament.render_options.xOffset -= (left_shift + 2);
+        }
+        if (ornament.render_options.xOffset < 0) {
+          left_shift += reportedWidth;
+        } else if (ornament.render_options.xOffset > 0) {
+          right_shift += reportedWidth;
+        }
 
-      if (ornament.getPosition() === Modifier.Position.ABOVE) {
-        ornament.setTextLine(state.top_text_line);
-        state.top_text_line += increment;
-      } else {
-        ornament.setTextLine(state.text_line);
-        state.text_line += increment;
+        // articulations above/below the line can be stacked.
+        if (Ornament.ornamentArticulation.indexOf(ornament.type) >= 0) {
+          // Unfortunately we con't know the stem direction.  So we base it
+          // on the line number, but also allow it to be overridden.
+          if (ornament.note.getLineNumber() >= 3 || ornament.getPosition() === Modifier.Position.ABOVE) {
+            state.top_text_line += increment;
+            ornament.render_options.yOffset += yOffset;
```

Comment:
Ornaments derive from `Modifier` which already have `x_shift` and `y_shift`, and you can call `setXShift/setYShift`, would those work here instead of adding custom `render_options`?